### ISSUE TITLE
Rewording 2.8.8 (time-based multi-factor OTP token)

### ISSUE THREAD

**Issue by randomstuff** [OWASP/ASVS#2462]
Current 2.8.8:

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **2.8.8** | [ADDED] Ensure that generation of the time-based multi-factor OTP token is based on the server's system time and not the client's machine. | | | ✓ | 367 |

Some things are off for me:

* "generation" should probably "verification" instead (the client should still use the client time for generating the token, right?)
* ensure → verify

Proposition:

> Ensure that verification of the time-based multi-factor OTP token is based on the verifier server's system time and not the authenticating client's machine.

**Comment by tghosth** [comment]
@randomstuff please review the previous discussion to confirm that your suggestion is correct:
https://github.com/OWASP/ASVS/issues/653#issuecomment-500548191

I think it is but I'd like to be sure

**Comment by elarlang** [comment]
The beginning of the requirement need to be changed from "Ensure" to "Verify" anyway.

**Comment by randomstuff** [comment]
Woops, I forgot about this one!

The requirement is:

> [ADDED] Ensure that generation of the time-based multi-factor OTP token is based on the server's system time and not the client's machine.

I think we are talking about TOTP-style authentication.

The client-side OTP generation component is usually completely disconnected from the target server. I could have a smartphone-based application on a smartphone which does not event have access to internet. It could be a hardware token without any network connectivity at all. In this context, requiring that the client-side generation of the OTP is based on the server's system time does not make sense.

The server-side OTP validation component on the other hand should use the server system time and not blindly trust some client-side time. Is this a problem in practice, however? Does this happen at all? As before, the client-side TOTP generation component is usually not the actual client so it would not make sense to use the client time at all as the client and the TOTP generation component might be on completely different devices.

**Comment by randomstuff** [comment]
Therefore I am wondering whether this requirement is really useful.

**Comment by tghosth** [comment]
@randomstuff did you read the previous discussion? https://github.com/OWASP/ASVS/issues/653#issuecomment-500548191

**Comment by randomstuff** [comment]
@tghosth. Yes, I read it. https://github.com/OWASP/ASVS/issues/653#issuecomment-500598195 mentions a case where this happened. I still have hard time really understanding this, especially in the (nominal) case where the OTP generation component and the client are unrelated  components.

**Comment by tghosth** [comment]
I think this is still a relevant issue based on the example in that comment. I would suggest:

> Verify that time-based OTPs are checked based on a time source from a trusted service and not from an untrusted or client provided time.

What do you think @randomstuff ?

**Comment by randomstuff** [comment]
Yes I think that's good. Is this clear enough that we are talking about the verifying side, not the providing side? (I *think* it is.)

**Comment by tghosth** [comment]
Thanks, @randomstuff I opened #2526
